### PR TITLE
Template API: Only generate templates on WCA routes

### PR DESCRIPTION
**Pull request by mattsherman**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In the review of #41316 we discovered that we were generating the product editor templates on every page request, including the front end and non-WCA admin pages.

This is wasteful (consuming CPU cycles for nothing) and also had the side-effect of quickly filling up the log file needlessly.

This PR addresses this by only generating the templates for WCA route requests.

#41316 will further improve matters by only logging template events when the template has actually changed.

### How to test the changes in this Pull Request:

Add the following template modifications to a `mu-plugin`:

```php
use Automattic\WooCommerce\Admin\BlockTemplates\BlockInterface;
use Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates\SectionInterface;

if ( ! function_exists( 'example_hook_up_block_template_modifications' ) ) {
	/**
	 * Hook up the block template after hooks.
	 */
	function example_hook_up_block_template_modifications() {
		/*
		 * Events that will be logged:
		 * - Block removed from template - INFO
		 */
		add_action(
			'woocommerce_block_template_area_product-form_after_add_block_product-summary',
			function ( BlockInterface $product_summary_field ) {
				$product_summary_field->remove();
			}
		);

	}
}

add_action( 'init', 'example_hook_up_block_template_modifications', 0 );
```

1. Enable the **Product Editor** under **WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**.
2. Go to **Products** > **Add New**
3. Verify the default logging threshold (`warning`) -- there should be 1 events
    - Look at the log (**WooCommerce** > **Status** > **Logs** > `block_template` files (multiple template generations will be in a single file, so look closely at the timestamps)
4. Reload the **Add New** page... there should be another log entry.
5. Reload the admin dashboard page (`/wp-admin/index.php`)... there should not be another log entry.
6. Reload other WCA pages, like the WooCommerce Home page (`/wp-admin/admin.php?page=wc-admin`)... there should be a new log entry for every reload. Note: if just navigating between WCA pages in the client, there will be no additional log entries, as the routing is done client-side.
7. Reload front-end pages. There should be no new log entries.

Regression testing:

8. Verify saving a new product in the editor works.
9. Verify creating a variable product with variations in the editor works.
10. Disable the product editor.
11. Verify the classic experience works (simple saving of a new product should suffice).

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php [hunk 1]
```diff
@@ -78,21 +72,9 @@ public function enqueue_scripts() {
...
 			return;
 		}
-		$post_type_object     = get_post_type_object( 'product' );
-		$block_editor_context = new WP_Block_Editor_Context( array( 'name' => self::EDITOR_CONTEXT_NAME ) );
-		$template_registry    = wc_get_container()->get( BlockTemplateRegistry::class );
 
-		$editor_settings = array();
-		if ( ! empty( $post_type_object->template ) ) {
-			$editor_settings['template']     = $post_type_object->template;
-			$editor_settings['templateLock'] = ! empty( $post_type_object->template_lock ) ? $post_type_object->template_lock : false;
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php:88]: This line is still necessary within your new `get_product_editor_settings()` as the block picker started showing up in the variation template, I am not sure why not in the simple product template. Maybe we do have template lock hardcoded for that template within the JS somewhere?
Screenshot 2023-11-14 at 11 50 29 AM

**Review by louwie17 — CHANGES_REQUESTED** [review]
Nice work @mattsherman, overall this tested great and the code looks good!
I did find one bug with the variation template, that showed up with some of the refactor work, see inline comment (should be a simple fix ).

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php (hunk 1 above)
Comment (mattsherman) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php:88]: Good catch. I'm going to fix this in the JS, as we always want `templateLock` set to `all`. We shouldn't be relying on the PHP to set this properly.

**Review by louwie17 — APPROVED** [review]
Thanks for the update @mattsherman, this looks great to me, nice work!!

**Pull request merged**